fuzz/fuzz_targets/fuzz.rs

```diff
@@ -0,0 +1,320 @@
+}
+
+/// An arbitrary mutation to apply to the BVH to fuzz BVH optimization.
+#[derive(Debug, Arbitrary)]
+enum ArbitraryMutation<const D: usize> {
+    Remove(usize),
+    Add(ArbitraryShape<D>),
+}
+
+#[derive(Copy, Clone, Debug, Arbitrary)]
+/// Controls whether the input is modified to help test certain properties.
+enum Mode {
+    /// AABB's may have mostly arbitrary bounds, and ray may have mostly arbitrary
+    /// origin and direction.
+    Chaos,
+    /// AABB's are unit cubes, and must have integer coordinates. Ray must have an
+    /// origin consisting of integer coordinates and a direction that is parallel to
+    /// one of the axes.
+    ///
+    /// In this mode, all types of traversal are expected to yield the same results,
+    /// except when bugs exist that have yet to be fixed.
+    Grid,
+}
+
+impl Mode {
+    fn is_grid(self) -> bool {
+        matches!(self, Self::Grid)
+    }
+}
+
+/// The complete set of inputs for a single fuzz iteration.
+#[derive(Debug, Arbitrary)]
+struct Workload<const D: usize> {
+    shapes: Vec<ArbitraryShape<D>>,
+    ray: ArbitraryRay<D>,
+    mutations: Vec<ArbitraryMutation<D>>,
+}
+
+impl<const D: usize> Workload<D> {
+    /// Called directly from the `cargo fuzz` entry point. Code in this function is
+    /// easier for `rust-analyzer`` than code in that macro.
+    fn fuzz(mut self) {
+        let mut bvh = Bvh::build(&mut self.shapes);
+        let ray = self.ray.ray();
+
+        if self.shapes.len()
+            + self
+                .mutations
+                .iter()
+                .filter(|m| matches!(m, ArbitraryMutation::Add(_)))
+                .count()
+            > 32
+        {
+            // Prevent traversal stack overflow by limiting max BVH depth to the traversal
+            // stack size limit.
+            return;
+        }
+
+        loop {
+            let assert_traversal_agreement =
+                self.ray.mode.is_grid() && self.shapes.iter().all(|s| s.mode.is_grid());
+
+            // Check that these don't panic.
+            bvh.assert_consistent(&self.shapes);
+            bvh.assert_tight();
+            let flat_bvh = bvh.flatten();
+
+            let traverse = bvh
+                .traverse(&ray, &self.shapes)
+                .into_iter()
+                .map(ByPtr)
+                .collect::<HashSet<_>>();
+            let traverse_iterator = bvh
+                .traverse_iterator(&ray, &self.shapes)
+                .map(ByPtr)
+                .collect::<HashSet<_>>();
+            let _traverse_flat = flat_bvh
+                .traverse(&ray, &self.shapes)
+                .into_iter()
+                .map(ByPtr)
+                .collect::<HashSet<_>>();
+
+            if assert_traversal_agreement {
+                assert_eq!(traverse, traverse_iterator);
+                // Fails, due to a bug.
+                // assert_eq!(traverse, traverse_flat);
+            } else {
+                // Fails, probably due to rounding errors.
```

Comment:
Let's add TODOs for the failures. Perhaps even add an additional tracking issue to look into these?

Reply:
Done and done